src/jv_aux.c

```diff
@@ -38,29 +47,39 @@ static int parse_slice(jv j, jv slice, int* pstart, int* pend) {
       jv_get_kind(end_jv) != JV_KIND_NUMBER) {
     jv_free(start_jv);
     jv_free(end_jv);
-    return 0;
-  } else {
-    double dstart = jv_number_value(start_jv);
-    double dend = jv_number_value(end_jv);
-    jv_free(start_jv);
-    jv_free(end_jv);
-    if (dstart < 0) dstart += len;
-    if (dend < 0) dend += len;
-    if (dstart < 0) dstart = 0;
-    if (dstart > len) dstart = len;
-
-    int start = (int)dstart;
-    int end = (dend > len) ? len : (int)dend;
-    // Ends are exclusive but e.g. 1 < 1.5 so :1.5 should be :2 not :1
-    if(end < dend) end += 1;
-
-    if (end > len) end = len;
-    if (end < start) end = start;
-    assert(0 <= start && start <= end && end <= len);
-    *pstart = start;
-    *pend = end;
-    return 1;
-  }
+    return jv_invalid_with_msg(jv_string("Array/string slice indices must be integers"));
+  }
+  double dstart = jv_number_value(start_jv);
+  double dend = jv_number_value(end_jv);
+  jv_free(start_jv);
+  jv_free(end_jv);
+  if (dstart < 0) dstart += len;
+  if (dend < 0) dend += len;
+  if (dstart < 0) dstart = 0;
+  if (dstart > len) dstart = len;
```

Comment:
If you clamp indices into `[0, len]`, you don't need to compare against integer min/max. I think we can simplify as follows.
```c
if (dstart < 0) dstart += len;
if (dstart < 0 || isnan(dstart))
  start = 0;
else if (dstart > len)
  start = len;
else
  start = (int)dstart;

if (dend < 0) dend += len;
if (dend > len || isnan(dend))
  end = len;
else if (dend < start)
  end = start;
else if ((int)dend == dend)
  end = (int)dend;
else
  end = (int)dend + 1;
```

Reply:
```C
if (dstart < 0 || isnan(dstart))
  start = 0;
```
won't do if we want NaN indices to yield `null` as in 1.6.

Reply:
But for slicing,
```sh
 $ jq -n '[1,2,3] | .[nan:1]'
[
  1
]
```

Reply:
This behavior might depend on an undefined behavior though. My comment above deals `nan` as `null` i.e. `.[nan:N] == `.[:N]` and `.[N:nan] == .[N:]`.

Reply:
I see jq 1.6 trip an assertion when the start index of an array slice is a NaN:
```
: ; jq-1.6 -cn '[1,2,3] | .[nan:1]'
jq-1.6: src/jv_aux.c:47: parse_slice: Assertion `0 <= start && start <= end && end <= len' failed.
Abort
```

But not for the end index:
```
: ; jq-1.6 -cn '[1,2,3] | .[1:nan]'
[]
```

So slicing with NaNs was UB in 1.6.